File: lambda-functions/get-assigned-candidates.py
import json
import boto3

# Initialize DynamoDB client
dynamodb = boto3.resource('dynamodb')

# TODO: Replace with your actual table names
CANDIDATES_TABLE_NAME = 'your-candidates-table-name'
CATEGORIZER_ASSIGNMENTS_TABLE = 'categorizer-assignments'

candidates_table = dynamodb.Table(CANDIDATES_TABLE_NAME)
assignments_table = dynamodb.Table(CATEGORIZER_ASSIGNMENTS_TABLE)

def lambda_handler(event, context):
    """
    Lambda function to get candidates assigned to a specific categorizer
    
    Query parameters:
    - categorizerId: The categorizer ID (e.g., CAT001)
    
    Example: GET /get-assigned-candidates?categorizerId=CAT001
    """
    
    try:
        # Get categorizer ID from query parameters
        categorizer_id = event.get('queryStringParameters', {}).get('categorizerId')
        
        if not categorizer_id:
            return {
                'statusCode': 400,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*',
                    'Access-Control-Allow-Headers': 'Content-Type',
                    'Access-Control-Allow-Methods': 'GET, OPTIONS'
                },
                'body': json.dumps({
                    'success': False,
                    'message': 'categorizerId query parameter is required'
                })
            }
        
        # Query assignments table for this categorizer
        # This assumes GSI on categorizerId or you scan with filter
        response = assignments_table.scan(
            FilterExpression='categorizerId = :categorizer',
            ExpressionAttributeValues={
                ':categorizer': categorizer_id
            }
        )
        
        assignment_items = response.get('Items', [])
        registration_ids = [item['pk'] for item in assignment_items]
        
        # Fetch full candidate details from main table
        candidates = []
        for reg_id in registration_ids:
            try:
                candidate_response = candidates_table.get_item(
                    Key={
                        'pk': reg_id,
                        # 'sk': 'CANDIDATE'  # Add if you use sort key
                    }
                )
                
                if 'Item' in candidate_response:
                    candidates.append(candidate_response['Item'])
                    
            except Exception as e:
                print(f"Error fetching candidate {reg_id}: {str(e)}")
        
        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Headers': 'Content-Type',
                'Access-Control-Allow-Methods': 'GET, OPTIONS'
            },
            'body': json.dumps(candidates, default=str)
        }
        
    except Exception as e:
        print(f"Error getting assigned candidates: {str(e)}")
        return {
            'statusCode': 500,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Headers': 'Content-Type',
                'Access-Control-Allow-Methods': 'GET, OPTIONS'
            },
            'body': json.dumps({
                'success': False,
                'message': f'Error fetching candidates: {str(e)}'
            })
        }

File: lambda-functions/get-assigned-candidates.js
// AWS Lambda Function - Get Assigned Candidates (Node.js)
// Matches your existing eventregistrations table structure
const { DynamoDBClient, ScanCommand } = require("@aws-sdk/client-dynamodb");
const { DynamoDBDocumentClient } = require("@aws-sdk/lib-dynamodb");

// Initialize DynamoDB client
const REGION = "ap-south-1";
const client = new DynamoDBClient({ region: REGION });
const docClient = DynamoDBDocumentClient.from(client);

// Your existing table name
const TABLE_NAME = 'eventregistrations';
// New table for categorizer assignments
const CATEGORIZER_ASSIGNMENTS_TABLE = 'categorizer-assignments';

const corsHeaders = {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Access-Control-Allow-Methods': 'GET, OPTIONS'
};

// Helper function to convert DynamoDB item to plain object
function unmarshallItem(item) {
    const result = {};
    for (const [key, value] of Object.entries(item)) {
        if (value.S) result[key] = value.S;
        else if (value.N) result[key] = value.N;
        else if (value.BOOL !== undefined) result[key] = value.BOOL;
        else if (value.NULL) result[key] = null;
        else result[key] = value;
    }
    return result;
}

exports.handler = async (event) => {
    try {
        // Handle OPTIONS for CORS preflight
        if (event.requestContext?.http?.method === 'OPTIONS' || event.httpMethod === 'OPTIONS') {
            return {
                statusCode: 200,
                headers: corsHeaders,
                body: ''
            };
        }

        // Get categorizer ID from query parameters
        const categorizerId = event.queryStringParameters?.categorizerId;

        if (!categorizerId) {
            return {
                statusCode: 400,
                headers: corsHeaders,
                body: JSON.stringify({
                    success: false,
                    message: 'categorizerId query parameter is required'
                })
            };
        }

        // 1. Get all assigned registration IDs for this categorizer
        const scanAssignmentsCommand = new ScanCommand({
            TableName: CATEGORIZER_ASSIGNMENTS_TABLE,
            FilterExpression: 'categorizerId = :categorizer',
            ExpressionAttributeValues: {
                ':categorizer': { S: categorizerId }
            }
        });

        const assignmentsResponse = await client.send(scanAssignmentsCommand);
        const assignmentItems = assignmentsResponse.Items || [];
        const assignedRegistrationIds = new Set(assignmentItems.map(item => item.pk?.S).filter(Boolean));

        if (assignedRegistrationIds.size === 0) {
            return {
                statusCode: 200,
                headers: corsHeaders,
                body: JSON.stringify([])
            };
        }

        console.log(`✅ Found ${assignedRegistrationIds.size} assignments for ${categorizerId}`);

        // 2. Scan the MAIN table once (Optimized: O(1) Scan vs O(N) Scans)
        // Since we don't have the PK/SK of the main table, we must scan it.
        // Scanning once is faster than scanning N times for N candidates.
        const scanMainTableCommand = new ScanCommand({
            TableName: TABLE_NAME
        });

        const mainTableResponse = await client.send(scanMainTableCommand);
        const allCandidates = mainTableResponse.Items || [];

        // 3. Filter in memory
        const matchedCandidates = [];
        for (const item of allCandidates) {
            const candidate = unmarshallItem(item);
            if (candidate.registrationId && assignedRegistrationIds.has(candidate.registrationId)) {
                matchedCandidates.push(candidate);
            }
        }

        console.log(`✅ Matched ${matchedCandidates.length} candidates from main table`);

        const responseHeaders = {
            ...corsHeaders,
            'X-Debug-Assignments-Found': assignedRegistrationIds.size.toString(),
            'X-Debug-Candidates-Fetched': matchedCandidates.length.toString()
        };

        return {
            statusCode: 200,
            headers: responseHeaders,
            body: JSON.stringify(matchedCandidates)
        };

    } catch (error) {
        console.error('❌ Error getting assigned candidates:', error);
        return {
            statusCode: 500,
            headers: corsHeaders,
            body: JSON.stringify({
                success: false,
                message: `Error fetching candidates: ${error.message}`
            })
        };
    }
};
